### PR TITLE
appropriate logic in cascade classifier when rejectLevels is true (#5347)

### PR DESCRIPTION
**Pull request by Dikay900**
Please see #5347 for more information.

### CODE REVIEW

**Comment by Dikay900** [comment]
I added the variable as requested please review if the name for the variable/parameter is appropriate and if the method for comparing the variable is ok or if it would be better to specify one specific stage.
@headupinclouds

**Comment by headupinclouds** [comment]
Thanks for adding this.  This looks good to me, but buildbot doesn't like it  I am unsure how strictly the ABI tests are followed wrt the default parameter issues it reports:

http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/4291/steps/Compare%20ABI%20dumps/logs/report-html

This seems like biggest potential issue:
```
The name of the appropriate symbol for this function on binary level has been changed from

_ZN2cv17CascadeClassifier16detectMultiScaleERKNS_11_InputArrayERSt6vectorINS_5Rect_IiEESaIS6_EERS4_IiSaIiEERS4_IdSaIdEEdiiNS_5Size_IiEESH_b 
to 
_ZN2cv17CascadeClassifier16detectMultiScaleERKNS_11_InputArrayERSt6vectorINS_5Rect_IiEESaIS6_EERS4_IiSaIiEERS4_IdSaIdEEdiiNS_5Size_IiEESH_bi. 
This may cause "undefined reference" linker error in old client applications.
```

I suppose the alternative could be to hard code the old behavior with the exact same prototype and add a new method to provide the flexibility of your PR.  Thoughts?  Happy to help.

File: modules/objdetect/include/opencv2/objdetect.hpp [hunk 1]
```diff
@@ -304,7 +305,8 @@ class CV_EXPORTS_W CascadeClassifier
                                   int minNeighbors = 3, int flags = 0,
                                   Size minSize = Size(),
                                   Size maxSize = Size(),
-                                  bool outputRejectLevels = false );
+                                  bool outputRejectLevels = false,
+                                  int rejectStage = 3);
```
Comment (StevenPuttemans) [inline modules/objdetect/include/opencv2/objdetect.hpp]: I would put the default value at 0, not 3. It is not the expected behaviour to output detections of the last 3 stages ... not if you do not want this... So why not set it to 0?

**Comment by StevenPuttemans** [comment]
Could you tell me a good reason why you would want to have the opportunity of visualising the last x stages of detections? Shouldnt people just train a model with less stages then if they want to have those detections?

File: modules/objdetect/include/opencv2/objdetect.hpp (hunk 1 above)
Comment (Dikay900) [inline modules/objdetect/include/opencv2/objdetect.hpp]: i set this to 3 that the previous behaviour before 0bc6092 is restored with the changed logic now

**Comment by Dikay900** [comment]
@headupinclouds could answer to this question.

For the ABI compatibility it seems that we cannot change it like this at all, so the only chance we got is maybe adding an additional method call but this seems a bit odd after already having 3 different method calls right now.

File: modules/objdetect/include/opencv2/objdetect.hpp (hunk 1 above)
Comment (StevenPuttemans) [inline modules/objdetect/include/opencv2/objdetect.hpp]: yeah but that behaviour was bogus .. it is not natural... so I would not do that.

Comment (Dikay900) [inline modules/objdetect/include/opencv2/objdetect.hpp]: Agreed and noted.
I will change this to default 0 in a future commit.

**Comment by headupinclouds** [comment]
> Shouldnt people just train a model with less stages then if they want to have those detections

I think if all detections are treated equally (regardless of rejection stage) that is true.  Hard coding the last four stages does seem very arbitrary.  Maybe this behavior isn't needed.

**Comment by Dikay900** [comment]
So we do not need this change right?
I did not think of training the model with less stages gives exactly the same effect.

**Comment by Dikay900** [comment]
So then the best functionality we got here is the one already implemented right?

**Comment by StevenPuttemans** [comment]
The one already implemented is the hardcoded `=0` one right? If so, yes that seems the best solution for me. Then I would add another overload function where you are able to specify which stage you want.

**Comment by Dikay900** [comment]
well unfortunately we have a problem here since the overloading is already at it's limit.
http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/4301/steps/compile%20release/logs/stdio

I cannot arrange another int like parameter in the default set parameter list which is similiar to the already existing method.
Any suggestions here besides creating a new method?

**Comment by Dikay900** [comment]
I hope this build and tests are correct now then please review @StevenPuttemans 
@headupinclouds please look as well if it is like you are expecting it.

**Comment by headupinclouds** [comment]
Thanks.  I will have to look on Monday.

**Comment by Dikay900** [comment]
seems like we cannot get arround this or am i doing something wrong?
http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/4319/steps/Compare%20ABI%20dumps/logs/report-html#Type_Binary_Problems_High

File: modules/objdetect/include/opencv2/objdetect.hpp [hunk 2]
```diff
@@ -199,6 +199,15 @@ class CV_EXPORTS_W BaseCascadeClassifier : public Algorithm
...
 
+    virtual void detectMultiScale( InputArray image,
+                                   CV_OUT std::vector<Rect>& objects,
+                                   CV_OUT std::vector<int>& rejectLevels,
+                                   CV_OUT std::vector<double>& levelWeights,
+                                   int rejectStage,
+                                   double scaleFactor,
+                                   int minNeighbors, int flags,
+                                   Size minSize, Size maxSize ) = 0;
+
```
Comment (alalek) [inline modules/objdetect/include/opencv2/objdetect.hpp]: To save ABI compatibility you need to move new **virtual** method to the end of virtual method list.

Also please add some TODO comment, like `OPENCV_ABI_COMPATIBILITY` mark (to move this method back later)

**Comment by StevenPuttemans** [comment]
quite busy at the moment but will have a look later today :)

File: modules/objdetect/include/opencv2/objdetect.hpp (hunk 2 above)
Comment (Dikay900) [inline modules/objdetect/include/opencv2/objdetect.hpp]: ok will do that, thanks for that hint

Comment (Dikay900) [inline modules/objdetect/include/opencv2/objdetect.hpp]: just another question: we are free to move the virtual method in another PR after this got added?

Comment (Dikay900) [inline modules/objdetect/include/opencv2/objdetect.hpp]: Did as you say but still ABI Compatibility still fails :/

**Comment by achalddave** [comment]
Looking back at this, the current check of
`if( classifier->data.stages.size() + result == 0 )`

also doesn't really do the right thing as it suffers from an off-by-one error (my fault); it should (I believe) fill `rejectLevels` and `levelWeights` with information about any proposal that failed the last stage. Instead, it fills them with information about proposals that passed the last stage (that is, that successfully passed the entire cascade).

**Comment by Dikay900** [comment]
I understand your concerns and agree that the current behaviour is not what you tried to explain.

For this PR i would like to get the parameter and the new method working and do a follow up PR which will change the real algorithm.

**Comment by StevenPuttemans** [comment]
I agee with @Dikay900 that it might be good to first unify the behaviour in between branches to avoid different end results and then try to find a solution for what is the best way of adapting the algorith.

As to the parameter name, lastAcceptedStage seems fine to me, indeed we do need a good documentation on its use then!

**Comment by Dikay900** [comment]
@StevenPuttemans i won't revert the behavior for this method to 2.4 for just one commit or a very specific commit range since we will change this anyway to another method and i rather want to have the same functionality in 3.0 if someone used it already till now.

Renamed the variable added TODO comment for changing this method in follow up PR (could be done by someone else)
And testing if i get around this ABI compatibility...

**Comment by StevenPuttemans** [comment]
:+1:

**Comment by Dikay900** [comment]
ok @achaldave i will change this line to
>-result >= lastAcceptedStage
with a later commit i want to see if ABI_compatibility will get through now.

**Comment by Dikay900** [comment]
have to call for help again @alalek
I still got the same errors as in the beginning
http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/4364/steps/Compare%20ABI%20dumps/logs/report-html#Type_Binary_Problems_High

**Comment by alalek** [comment]
@Dikay900 Looks like there is no easy way to resolve ABI problem with virtual functions.

**Comment by Dikay900** [comment]
Then it seems we cannot change anything in the CascadeClassifier without hacks or mess up the structure since the method calls goes back to the virtual methods of BaseCascadeClassifier.

**Comment by achalddave** [comment]
Perhaps we can maintain the old, 2.0 version of the code in the default method, deprecate it in documentation, and add a new method that uses the new method?

**Comment by Dikay900** [comment]
unfortunately we cannot change any parameter in the base function you will call when you use the cascade so the only methods we can use are the three we currently have. We do not have any method with a parameter to provide this functionality we want, since we only can set outputRejectLevels to true or false.

**Comment by headupinclouds** [comment]
#define CV_HAAR_DO_CANNY_PRUNING    1
#define CV_HAAR_SCALE_IMAGE         2
#define CV_HAAR_FIND_BIGGEST_OBJECT 4
#define CV_HAAR_DO_ROUGH_SEARCH     8

**Comment by Dikay900** [comment]
well i really do not like that idea although this is seems to be the only way to do this.
But i do not want to include such a workaround, i will change this PR to reflect the changed if line with default value 0 since no default value will reflect the old 2.4 behavior.

**Comment by vpisarev** [comment]
:+1:

**Comment by Dikay900** [comment]
will get reverted until ABI_Compatibility is resolved.
I think the current structure of the refactored Cascadeclassifier results into the diamond problem so we cannot get around the abi compiler it seems. (was my last thought digging through the code)
